File: strings/06_palindrome/palindrome.py
class Solution:
    def is_palindrome(self, x: int) -> bool:
        if not self._check_constraints(x):
            raise ValueError("Constraints violated")
        if x < 0:
            return False

        x_reversed = 0
        tmp = x
        while tmp != 0:
            digit = tmp % 10
            x_reversed = x_reversed * 10 + digit
            tmp //= 10

        return x_reversed == x

    def _check_constraints(self, x: int) -> bool:
        if x < -2**31 or x > 2**31 - 1:
            return False

        return True

File: strings/06_palindrome/palindrome.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <stdexcept>
#include <cstdint>

using namespace std;


class Solution {
public:
    bool isPalindrome(int x) {
        if (!checkConstraints(x)) {
            throw std::runtime_error("Constraints violated");
        }
        if (x < 0) return false;

        int reversed = 0;
        int tmp = x;
        while (tmp != 0) {
            int digit = tmp % 10;
            reversed = reversed * 10 + digit;
            tmp = tmp / 10;
        }
        return (reversed == x);
    }

private:
    // Helper function to check the constraints
    bool checkConstraints(const int& x) {
        if (x < pow(-2, 31) || x > pow(2, 31) - 1) return false;

        return true;
    }
};
